### DISCUSSION TITLE
Cannot figure out how to start KDE properly under xrdp.

### DISCUSSION THREAD

**Discussion by allefeld** [neutrinolabs/xrdp#2126] [Q&A]
I'm trying to start my standard KDE session under xrdp.
After a lot of googling, one hint seemed to work: put `/usr/bin/startkde` into `~/.Xclients`.

Now KDE starts and is mostly okay, but I noticed some weirdness. For example, normally `~/.bash_profile` has been executed somewhere during KDE start, so changes to `PATH` I made there are effective, e.g. inherited by shells running in a terminal window (Konsole). Since this is different, I'm wondering whether other things are, too.

What I would like to achieve is that xrdp starts a desktop session in the exact same way as a display manager like sddm does. Is that possible?

In case that is relevant, the computer runs Debian 10, with the included KDE Plasma 5.14.5 and xrdp 0.9.9.

**Comment by matt335672** [comment]
There will differences from a local session for two main reasons:-
1) The PATH is different on your version of xrdp . For more details see neutrinolabs/xrdp#1773 from around [this post](/neutrinolabs/xrdp/issues/1773#issuecomment-758571408), and the consequent Debian bug which was raised https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=984782. This is a Debian packaging issue, but you can work around it by making changes to `/etc/xrdp/startwm.sh`.
2) The permissions you get will be different because of a system feature called 'polkit'. Essentially, polkit realises you're in a remote session and limits what you can do. Here's a [list](/neutrinolabs/xrdp/issues?q=is%3Aissue+is%3Aclosed+polkit+0.105) of the issues raised by other users related to this feature. You'll need to work out which features you want to allow in your remote session and add polkit actions to enable them for 'implicit any' sessions.

**Reply by allefeld** [reply]
Thanks! I did not expect it to be so complex.

So far, the PATH issue is the only one I'm experiencing, so I'll focus on that.
For example I do have "reboot" and "shutdown" options in KDE.

**Reply by allefeld** [reply]
Regarding PATH, I'm not sure my problem is exactly the same as in the issue you linked to. I don't have anything in `/etc/environment/`, my problem is about my user's local additions to PATH made in my `~/.bash_profile`.

Do I have to do something like mentioned [here](https://github.com/neutrinolabs/xrdp/issues/1130#issuecomment-390498948)?

**Reply by allefeld** [reply]
Assuming that `~/.Xclients` is a shell script, I changed it to this:

````
. "/home/myuser/.bash_profile"
/usr/bin/startkde
````

No change. KDE still starts, but `~/.bash_profile` does not seem to get executed.

Moreover, I could find almost no information on `~/.Xclients`. Why are there so many different files involved in starting X anyway?!

**Reply by matt335672** [reply]
Normally the session start called `/etc/xrdp/startwm.sh` which should be calling directly into your `.bash_profile`. If that's not working, there may be something in your `.bash_profile` which is causing the PATH to not be set.

Try this on the command line so you can see what happens when the file gets scanned:-

```
set -x
. ~/.bash_profile
set +x
```

I feel your pain regarding X startup. It's grown up on many systems over many years, so it's a bit of a mess.

**Reply by allefeld** [reply]
There's a lot of output. Most of it is from `~/.bashrc`, which is sourced in `~/.bash_profile`, and that in turn runs a script to activate powerline, which is a bash prompt beautifier.

I did the sourcing because to my knowledge that's the recommended thing; but now I'm wondering, there's really no need to do all the stuff in `~/.bashrc` on KDE start. ~~I guess I have to rethink my configuration.~~ – I remembered that my `~/.bashrc` contains lines that stop execution at a certain point for a non-interactive shell, so everything after `+++ LC_TIME=en_DK.UTF-8` doesn't happen.

Apart from that, as far as I can tell everything works properly:

````
+ . /home/ac/.bash_profile
++ PATH=/home/ac/bin:/home/ac/bin/xbin:/home/ac/.local/bin:/opt/conda/condabin:/home/ac/bin:/home/ac/bin/xbin:/home/ac/.local/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games:/snap/bin
++ XDG_DATA_DIRS=/home/ca/bin:/home/ca/bin:/home/ac/.local/share/flatpak/exports/share:/var/lib/flatpak/exports/share:/usr/local/share:/usr/share:/var/lib/snapd/desktop
++ . /home/ac/.bashrc
+++ umask u=rwx,g=rx,o=rx
+++ export LANG=en_US.UTF-8
+++ LANG=en_US.UTF-8
+++ export LC_TIME=en_DK.UTF-8
+++ LC_TIME=en_DK.UTF-8
+++ '[' -z '\[\] ac \[\] \[\]~ \[\] \[\]' ']'
+++ '[' 1000 -ne 0 ']'
+++ PS1='\u@\[\033[01;16m\]${HOSTNAME,,}\[\033[00m\]:\w\$ '
+++ shopt -s histappend
+++ bind '"\e[A": history-search-backward'
+++ bind '"\e[B": history-search-forward'
+++ bind 'set bind-tty-special-chars off'
+++ bind '"\C-u": kill-whole-line'
+++ shopt -s checkwinsize
+++ shopt -s globstar
++++ dircolors -b /home/ac/.dircolors
+++ eval 'LS_COLORS='\''no=00:fi=37:ln=37:or=37;41:di=32:pi=36:so=35:do=35:bd=33:cd=33:ex=31:su=37;41:sg=30;43:tw=30;42:ow=32:st=37;44:*.tar=01;32:*.tgz=01;32:*.taz=01;32:*.arj=01;32:*.lzh=01;32:*.zip=01;32:*.z=01;32:*.Z=01;32:*.gz=01;32:*.bz2=01;32:*.deb=01;32:*.rpm=01;32:*.jar=01;32:*.rar=01;32:*.thm=01;32:*.iso=01;32:*.dmg=01;32:*.jpg=01;35:*.jpeg=01;35:*.png=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.xcf=01;35:*.ico=01;35:*.mpg=01;36:*.mpeg=01;36:*.vob=01;36:*.mpv=01;36:*.dvd=01;36:*.rm=01;36:*.avi=01;36:*.wmv=01;36:*.mov=01;36:*.fli=01;36:*.flc=01;36:*.mng=01;36:*.3gp=01;36:*.ram=01;36:*.flv=01;36:*.mp4=01;36:*.mkv=01;36:*.wav=01;34:*.ogg=01;34:*.mp3=01;34:*.mp2=01;34:*.ra=01;34:*.wma=01;34:*.m4a=01;34:*.m3u=01;34:*.flac=01;34:*.ps=01;33:*.eps=01;33:*.ps.gz=01;33:*.pdf=01;33:*.doc=01;33:*.docx=01;33:*.odt=01;33:*.svg=01;33:*.cbz=01;33:*.cbr=01;33:*.epub=01;33:*.txt=01;37:*.c=01;37:*.h=01;37:*.m=01;37:*.pl=01;37:*.py=01;37:*.js=01;37:*.java=01;37:*.tex=01;31:*.bib=01;31:*.html=01;31:*.htm=01;31:*.css=01;31:*.xml=01;31:*.shtml=01;31:*.kml=01;31:*.gpx=01;31:*.kvtml=01;31:*.text=01;31:*.md.txt=01;31:*.md=01;31:*.blg=01;30:*.bbl=01;30:*.toc=01;30:*.log=01;30:*.lof=01;30:*.dvi=01;30:*.aux=01;30:*.out=01;30:*.o=01;30:'\'';
export LS_COLORS'
++++ LS_COLORS='no=00:fi=37:ln=37:or=37;41:di=32:pi=36:so=35:do=35:bd=33:cd=33:ex=31:su=37;41:sg=30;43:tw=30;42:ow=32:st=37;44:*.tar=01;32:*.tgz=01;32:*.taz=01;32:*.arj=01;32:*.lzh=01;32:*.zip=01;32:*.z=01;32:*.Z=01;32:*.gz=01;32:*.bz2=01;32:*.deb=01;32:*.rpm=01;32:*.jar=01;32:*.rar=01;32:*.thm=01;32:*.iso=01;32:*.dmg=01;32:*.jpg=01;35:*.jpeg=01;35:*.png=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.xcf=01;35:*.ico=01;35:*.mpg=01;36:*.mpeg=01;36:*.vob=01;36:*.mpv=01;36:*.dvd=01;36:*.rm=01;36:*.avi=01;36:*.wmv=01;36:*.mov=01;36:*.fli=01;36:*.flc=01;36:*.mng=01;36:*.3gp=01;36:*.ram=01;36:*.flv=01;36:*.mp4=01;36:*.mkv=01;36:*.wav=01;34:*.ogg=01;34:*.mp3=01;34:*.mp2=01;34:*.ra=01;34:*.wma=01;34:*.m4a=01;34:*.m3u=01;34:*.flac=01;34:*.ps=01;33:*.eps=01;33:*.ps.gz=01;33:*.pdf=01;33:*.doc=01;33:*.docx=01;33:*.odt=01;33:*.svg=01;33:*.cbz=01;33:*.cbr=01;33:*.epub=01;33:*.txt=01;37:*.c=01;37:*.h=01;37:*.m=01;37:*.pl=01;37:*.py=01;37:*.js=01;37:*.java=01;37:*.tex=01;31:*.bib=01;31:*.html=01;31:*.htm=01;31:*.css=01;31:*.xml=01;31:*.shtml=01;31:*.kml=01;31:*.gpx=01;31:*.kvtml=01;31:*.text=01;31:*.md.txt=01;31:*.md=01;31:*.blg=01;30:*.bbl=01;30:*.toc=01;30:*.log=01;30:*.lof=01;30:*.dvi=01;30:*.aux=01;30:*.out=01;30:*.o=01;30:'
++++ export LS_COLORS
+++ alias 'grep=grep --color=auto'
+++ alias 'fgrep=fgrep --color=auto'
+++ alias 'egrep=egrep --color=auto'
+++ alias 'ls=ls -lrtp --time-style +"%F %T" --color=auto'
+++ alias 'rm=rm -I'
++++ lessfile
+++ eval 'export LESSOPEN="/usr/bin/lessfile %s";
export LESSCLOSE="/usr/bin/lessfile %s %s";'
++++ export 'LESSOPEN=/usr/bin/lessfile %s'
++++ LESSOPEN='/usr/bin/lessfile %s'
++++ export 'LESSCLOSE=/usr/bin/lessfile %s %s'
++++ LESSCLOSE='/usr/bin/lessfile %s %s'
+++ export LESS=-R
+++ LESS=-R
+++ source /opt/conda/etc/profile.d/conda.sh
++++ export CONDA_EXE=/opt/conda/bin/conda
++++ CONDA_EXE=/opt/conda/bin/conda
++++ export _CE_M=
++++ _CE_M=
++++ export _CE_CONDA=
++++ _CE_CONDA=
++++ export CONDA_PYTHON_EXE=/opt/conda/bin/python
++++ CONDA_PYTHON_EXE=/opt/conda/bin/python
++++ '[' -z x ']'
+++ export ATOM_ACCESS_TOKEN=REDACTED
+++ ATOM_ACCESS_TOKEN=REDACTED
+++ powerline-daemon -q
+++ . /usr/share/powerline/bindings/bash/powerline.sh
+++++ _powerline_return 0
+++++ _powerline_return 43
+++++ return 0
+++++ return 43
+++++ test '0 43' = '0 43'
+++++ echo 0
++++ _POWERLINE_HAS_PIPESTATUS=0
++++ test -z powerline-config
++++ powerline-config shell --shell=bash uses prompt
++++ _powerline_setup_prompt
++++ VIRTUAL_ENV_DISABLE_PROMPT=1
++++ test -z powerline
++++ test '_powerline_status_wrapper _powerline_set_prompt
' '!=' '_powerline_status_wrapper '
+++++ _powerline_local_prompt left -r.bash 0 0 0 continuation
+++++ powerline shell left -r.bash --last-exit-code=0 --last-pipe-status=0 --jobnum=0 --renderer-arg=client_id=23104 --renderer-arg=local_theme=continuation
++++ PS2='\[\]        \[\] \[\]'
+++++ _powerline_local_prompt left '' 0 0 0 select
+++++ powerline shell left --last-exit-code=0 --last-pipe-status=0 --jobnum=0 --renderer-arg=client_id=23104 --renderer-arg=local_theme=select
++++ PS3=' Select variant  '
++++ powerline-config shell --shell=bash uses tmux
++++ _powerline_init_tmux_support
++++ test -n ''
+++ export ELECTRON_TRASH=gio
+++ ELECTRON_TRASH=gio
++ _powerline_status_wrapper _powerline_set_prompt
++ last_pipe_status=("${PIPESTATUS[@]}")
++ local last_exit_code=0 last_pipe_status
++ _powerline_has_pipestatus
++ return 0
++ test 1 -eq 0
++ test 0 '!=' 0
++ _powerline_set_prompt 0 0
++ local last_exit_code=0
++ shift
++ local last_pipe_status=0
++ shift
+++ wc -l
+++ jobs -p
++ local jobnum=0
+++ _powerline_prompt aboveleft 0 0 0
+++ powerline shell aboveleft --width=183 -r.bash --last-exit-code=0 --last-pipe-status=0 --jobnum=0 --renderer-arg=client_id=23104
++ PS1='\[\] ac \[\] \[\]~ \[\] \[\]'
++ test -n ''
++ test -n ''
++ return 0
````

**Reply by allefeld** [reply]
In other news, I switched from `~/.Xclients` to `~/.xsession`, and made that a proper bash script:

````
#!/bin/bash
. "$HOME/.bash_profile"
/usr/bin/startkde
````

That seems to work.

But I still don't understand why `startkde` doesn't execute `~/.bash_profile` by itself. Because it assumes that has already been done? When?

**Reply by matt335672** [reply]
startkde is normally executed from a shell script run in the user's context. The shell will be a non-interactive login shell.

A shell started within KDE will be an interactive non-login shell, so the files that get run are different.

Both of these situations are described in the [INVOCATION section of the bash manpage](https://man7.org/linux/man-pages/man1/bash.1.html#INVOCATION).

It's all rather confusing, and a lot of people just make .bashrc a symlink to .bash_profile (or vice-versa), and then use something like the following to decide what to run:-

```bash
if [[ $- = *i* ]]; then
    # Interactive shell
else
    # Non-interactive shell
fi
```

Go with whatever feels the most natural way for you.

**Reply by allefeld** [reply]
> startkde is normally executed from a shell script run in the user's context. The shell will be a non-interactive login shell.

Hm, I'm still not sure I understand: Does "normally" include execution by xrdp? If yes, meaning xrdp runs a non-interactive login shell from which startkde is executed, that would mean that that shell reads `~/.bash_profile` on startup, so that PATH etc. set there should be inherited by KDE. So why do I have to explicitly source it in my `~/.xsession`?

Thanks for the tip about `$- = *i*`! My `~/.bashrc` contains `[ -z "$PS1" ] && return` to stop processing for interactive shells, picked up at some point from somewhere, but your version is better.

**Reply by matt335672** [reply]
If you're on Debian, there's a problem with `/etc/xrdp/startwm.sh` not being run as a login shell (Fedora works around this differently). So although `/etc/profile` is scanned, `~/.bash_profile` won't be. I think that explains why you need to source it in your `~/.xsession`.

**Reply by allefeld** [reply]
Thanks for all your help!
